highfidelity/hifi
# Commerce: Support pagination at History endpoint; improve Purchases scrolling

This PR:
1. Adds client-side pagination (endless scrolling) support to the Recent Activity view in Wallet Home
    - The view will only auto-refresh if the user is looking at the top of the list. As soon as the user starts scrolling, the view will not auto-refresh.
2. Improves the scrolling behavior of Purchases, hopefully fixing [MS10296](https://highfidelity.fogbugz.com/f/cases/10296/My-Purchases-tab-in-marketplace-jumps-around-when-scrolled-through)

## Test Plan
1. IF YOU ALREADY HAVE at least 110 Recent Activity entries in Recent Activity, move ahead with Step 2. If you don't, switch to the Staging metaverse (see @conklin94122 if you need help there), then purchase enough items from the Marketplace to generate at least 110 Recent Activity entries - this means, for example, purchasing 55+ costed items, or 110 free items. _(I'm sorry, this is a pain, but there is no other way to test this PR)_
2. Open the WALLET app, and scroll all the way down in Recent Activity. Verify that the bottom-most entry is the one regarding your receipt of 10,000 HFC.
3. Close and re-open the WALLET app, and remain at the top of the Recent Activity list. Open your debug logs (using CTRL+SHIFT+L). Ensure the "DEBUG" and "INFO" checkboxes are checked. Verify that you see the logs `[INFO] [hifi.commerce] Sending "history"...` in there every four seconds.
4. Scroll down in Recent Activity so that you're not viewing the top of the list. In the debug logs, verify that you DO NOT see the logs `[INFO] [hifi.commerce] Sending "history"...` in there every four seconds.
5. Open My Purchases (via the Market app). Scroll up and down through My Purchases and verify that the scrolling behavior feels reasonable to you and that the ListView doesn't jump around.
6. Purchase a free item from the Marketplace. Once you see the purchase confirmation screen, immediately switch to My Purchases. Verify that you see an item at the top of the ListView that is marked as "pending".
7. Scroll down to the second or third item in the list, wait 10 seconds, and verify that the ListView is stable (i.e. the scrolling position doesn't jump around).
8. Scroll back up to the top of the list. If you don't see the top item marked as "pending", purchase another free item and then return to My Purchases, and then continue to Step 9.
9. Wait. Verify that, after some time, the top list item's status temporarily changes to "CONFIRMED!", and that the list doesn't auto-scroll.
